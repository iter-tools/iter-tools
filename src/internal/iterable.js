import { isIterable } from '../impls/is-iterable/is-iterable.js';
import { isLoopable } from '../impls/is-loopable/is-loopable.js';
import { isWrappable } from '../impls/is-wrappable/is-wrappable.js';
import { nullableWrap as wrap } from './wrap.js';
import { variadicCurryWithValidation } from './curry.js';
import { _, __iterate } from './symbols.js';

export { wrap, isIterable, isLoopable, isWrappable };

export function ensureIterable(value) {
  if (!isWrappable(value)) {
    if (typeof value.next === 'function') {
      throw new TypeError(
        'iter-tools received a value that looked like an iterator but was not iterable. Get help fixing this: https://github.com/iter-tools/iter-tools/wiki/Making-iterators-iterable',
      );
    } else throw new TypeError('Expected an iterable, null, or undefined');
  } else if (isIterable(value)) {
    return value;
  } else {
    return wrap(value);
  }
}

export function callReturn(iterator) {
  if ('return' in iterator) iterator.return();
}

export function BaseResultIterable(fn, args, iterablesArg) {
  this[_] = {
    fn,
    args,
    iterablesArg,
    staticIterator: null,
  };
}

Object.assign(BaseResultIterable.prototype, {
  constructor: BaseResultIterable,

  [__iterate]() {
    const { fn, iterablesArg, args } = this[_];
    return fn(iterablesArg, ...args);
  },

  next() {
    const this_ = this[_];
    this_.staticIterator = this_.staticIterator || this[__iterate]();
    return this_.staticIterator.next();
  },

  return(value) {
    const this_ = this[_];
    this_.staticIterator = this_.staticIterator || this[__iterate]();
    return this_.staticIterator.return(value);
  },

  throw(error) {
    const this_ = this[_];
    this_.staticIterator = this_.staticIterator || this[__iterate]();
    return this_.staticIterator.throw(error);
  },
});

export function ResultIterable(...args) {
  BaseResultIterable.apply(this, args);
}

ResultIterable.prototype = Object.assign(Object.create(BaseResultIterable.prototype), {
  constructor: ResultIterable,
  [Symbol.iterator]() {
    return this[__iterate]();
  },
});

function SimpleResultIterable(...args) {
  ResultIterable.apply(this, args);
}

SimpleResultIterable.prototype = Object.assign(Object.create(ResultIterable.prototype), {
  constructor: SimpleResultIterable,
  [__iterate]() {
    return this[_].fn(...this[_].args);
  },
});

function makeFunctionConfig(fn, fnConfig = {}) {
  const { validateArgs, variadic, reduces, optionalArgsAtEnd, minArgs, maxArgs } = fnConfig;

  return {
    fn,
    validateArgs: validateArgs || ((_) => {}),
    variadic: !!variadic,
    reduces: !!reduces,
    optionalArgsAtEnd: !!optionalArgsAtEnd,
    minArgs: minArgs === undefined ? fn.length - 1 : minArgs,
    maxArgs: maxArgs === undefined ? fn.length - 1 : maxArgs,
    isIterable: isWrappable,
    iterableType: 'iterable',
    applyOnIterableArgs: ensureIterable,
    IterableClass: ResultIterable,
  };
}

export function cache(it) {
  return [...it];
}

export function wrapWithResultIterable(fn, { validateArgs = (_) => _ } = {}) {
  return (...args) => {
    validateArgs(args);
    return new SimpleResultIterable(fn, args);
  };
}

export const iterableCurry = (fn, config) => {
  return variadicCurryWithValidation(makeFunctionConfig(fn, config));
};
