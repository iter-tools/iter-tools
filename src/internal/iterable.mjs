import { variadicCurryWithValidation } from './curry'
const emptyArr = []

export function isIterable (i) {
  return Boolean(i != null && i[Symbol.iterator])
}

export function ensureIterable (i) {
  if (i == null) {
    return emptyArr[Symbol.iterator]()
  } else if (!isIterable(i)) {
    if (typeof i.next === 'function') {
      throw new TypeError('Iterators are not supported arguments to iter-tools. You must wrap them using the `iterable` method.')
    }
    throw new TypeError('The argument is not an iterable or null')
  }
  return i
}

export function isValidIterableArgument (i) {
  return i == null || isIterable(i)
}

export const iterableCurry = (fn, { variadic = false, reduces = false } = {}, minArgs, maxArgs) => {
  return variadicCurryWithValidation(
    isValidIterableArgument,
    'iterable',
    ensureIterable,
    fn,
    variadic,
    reduces,
    false,
    minArgs,
    maxArgs
  )
}
