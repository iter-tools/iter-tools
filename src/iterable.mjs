const emptyArray = []

export default function iterable (iterator) {
  if (iterator == null) {
    return emptyArray
  } else if (iterator[Symbol.iterator]) {
    return iterator
  } else if (typeof iterator !== 'object' || typeof iterator.next !== 'function') {
    throw new Error(`Expected to receive an iterator of the form {next()}, but instead received: ${iterator}`)
  }

  return {
    [Symbol.iterator] () {
      return iterator
    }
  }
}
