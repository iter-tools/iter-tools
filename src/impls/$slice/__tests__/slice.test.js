/**
 * @generated-from ./$slice.test.js
 * This file is autogenerated from a template. Please do not edit it directly.
 * To rebuild it from its template use the command
 * > npm run generate
 * More information can be found in CONTRIBUTING.md
 */

import { slice } from 'iter-tools-es';
import { Iterable } from '../../../types/iterable.js';
import { wrap, unwrap } from '../../../test/helpers.js';

describe('slice', () => {
  let list: Iterable<number>;

  beforeEach(() => {
    list = wrap([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);
  });

  describe('with positional args', () => {
    it('can slice starting from the beginning', () => {
      expect(unwrap(slice(0, 2, list))).toEqual([0, 1]);
    });

    it('can slice to the end', () => {
      expect(unwrap(slice(8, list))).toEqual([8, 9]);
    });

    it('can slice with a step', () => {
      expect(unwrap(slice(2, 7, 2, list))).toEqual([2, 4, 6]);
    });
  });

  describe('with options argument', () => {
    it('can slice to the end', () => {
      expect(unwrap(slice({ start: 8 }, list))).toEqual([8, 9]);
    });

    it('returns simple slice with start/end', () => {
      expect(unwrap(slice({ start: 1, end: 4 }, list))).toEqual([1, 2, 3]);
    });

    it('returns empty slice with end > start', () => {
      expect(unwrap(slice({ start: 4, end: 3 }, list))).toEqual([]);
    });

    it('returns simple slice with start/end/step', () => {
      expect(unwrap(slice({ start: 1, end: 6, step: 2 }, list))).toEqual([1, 3, 5]);
    });

    it('returns empty iterable when passed out of bounds indicies', () => {
      expect(unwrap(slice({ start: 1, end: 6, step: 2 }, []))).toEqual([]);
    });

    it('returns empty iterable when passed null iterable', () => {
      expect(unwrap(slice({ start: 1, end: 4 }, null))).toEqual([]);
    });

    it('returns curried slice', () => {
      expect(unwrap(slice({ start: 0, end: 2 })(list))).toEqual([0, 1]);
    });

    it('returns slice with negative end', () => {
      expect(unwrap(slice({ start: 4, end: -2 }, list))).toEqual([4, 5, 6, 7]);
    });

    it('returns empty slice with negative end', () => {
      expect(unwrap(slice({ start: 4, end: -8 }, list))).toEqual([]);
    });

    it('returns slice with negative end and step', () => {
      expect(unwrap(slice({ start: 1, end: -1, step: 3 }, list))).toEqual([1, 4, 7]);
    });

    it('returns slice with negative start', () => {
      expect(unwrap(slice({ start: -3 }, list))).toEqual([7, 8, 9]);
    });

    it('returns slice with negative start and step', () => {
      expect(unwrap(slice({ start: -6, step: 2 }, list))).toEqual([4, 6, 8]);
    });

    it('returns slice with negative start and end', () => {
      expect(unwrap(slice({ start: -3, end: -1 }, list))).toEqual([7, 8]);
    });

    it('returns slice with negative start and end, and step', () => {
      expect(unwrap(slice({ start: -5, end: -1, step: 2 }, list))).toEqual([5, 7]);
    });

    it('returns slice with negative start and end, and end < start', () => {
      expect(unwrap(slice({ start: -1, end: -2 }, list))).toEqual([]);
    });

    it('returns slice with negative start and positive end', () => {
      expect(unwrap(slice({ start: -5, end: 9 }, list))).toEqual([5, 6, 7, 8]);
    });

    it('returns slice with negative start and positive end (return empty)', () => {
      expect(unwrap(slice({ start: -5, end: 2 }, list))).toEqual([]);
    });

    it('returns slice with negative start and positive end and step', () => {
      expect(unwrap(slice({ start: -5, end: 9, step: 2 }, list))).toEqual([5, 7]);
    });

    it('returns a smaller slice when start is negative and larger than source', () => {
      expect(unwrap(slice({ start: -3, end: 1, step: 1 }, [1, 2]))).toEqual([1]);
    });
  });

  it('when given invalid parameters', () => {
    const untypedSlice: any = slice;
    expect(() => untypedSlice('', [])).toThrowErrorMatchingSnapshot();
    expect(() => untypedSlice(0, {}, [])).toThrowErrorMatchingSnapshot();
    expect(() => untypedSlice(0, 0, -2, [])).toThrowErrorMatchingSnapshot();
  });
});
