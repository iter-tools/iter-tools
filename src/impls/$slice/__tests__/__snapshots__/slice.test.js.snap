// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`slice when given invalid parameters 1`] = `"start argument to slice must be a number"`;

exports[`slice when given invalid parameters 2`] = `"end argument to slice must be a number"`;

exports[`slice when given invalid parameters 3`] = `"step argument to slice must be a number > 0"`;
